core/shared/src/main/scala/zio/stm/TPriorityQueue.scala

```diff
@@ -0,0 +1,123 @@
+ * `Ordering` is defined on. Unlike a `TQueue`, `take` returns the highest
+ * priority value (the value that is first in the specified ordering) as
+ * opposed to the first value offered to the queue. The ordering that elements
+ * with the same priority will be taken from the queue is not guaranteed.
+ */
+final class TPriorityQueue[K, V] private (private val tref: TRef[SortedMap[K, ::[V]]]) extends AnyVal {
+
+  /**
+   * Offers the specified value to the queue with the specified priority.
+   */
+  def offer(key: K, value: V): USTM[Unit] =
+    tref.update { map =>
+      map.get(key) match {
+        case None         => map + (key -> ::(value, Nil))
+        case Some(values) => map + (key -> ::(value, values))
+      }
+    }
+
+  /**
+   * Offers all of the elements in the specified collection to the queue.
+   */
+  def offerAll(values: Iterable[(K, V)]): USTM[Unit] =
+    ZSTM.foreach_(values) { case (key, value) => offer(key, value) }
+
+  /**
+   * Peeks at the first value in the queue without removing it, retrying until
+   * a value is in the queue.
+   */
+  def peek: USTM[V] =
+    tref.get.flatMap { map =>
+      map.headOption match {
+        case None                  => ZSTM.retry
+        case Some((_, value :: _)) => ZSTM.succeedNow(value)
+      }
+    }
+
+  /**
+   * Peeks at the first value in the queue without removing it, returning
+   * `None` if there is not a value in the queue.
+   */
+  def peekOption: USTM[Option[V]] =
+    tref.get.map { map =>
+      map.headOption match {
+        case None                  => None
+        case Some((_, value :: _)) => Some(value)
+      }
+    }
+
+  /**
+   * Returns the size of the queue.
+   */
+  def size: USTM[Int] =
+    tref.get.map(_.values.map(_.length).sum)
+
+  /**
+   * Takes a value from the queue, retrying until a value is in the queue.
+   */
+  def take: USTM[V] =
+    tref.get.flatMap { map =>
+      map.headOption match {
+        case None                                      => ZSTM.retry
+        case Some((key, value :: (values @ ::(_, _)))) => tref.update(_ + (key -> values)).as(value)
+        case Some((key, value :: _))                   => tref.update(_ - key).as(value)
+      }
+    }
+
+  /**
+   * Takes all values from the queue.
+   */
+  def takeAll: USTM[List[V]] =
+    tref.modify(map => (map.values.flatten.toList, SortedMap.empty[K, ::[V]](map.ordering)))
+
+  /**
+   * Takes a value from the queue, returning `None` if there is not a value in
+   * the queue.
+   */
+  def takeOption: USTM[Option[V]] =
+    tref.get.flatMap { map =>
+      map.headOption match {
+        case None                                    => ZSTM.none
+        case Some((key, value :: Nil))               => tref.update(_ - key).as(Some(value))
+        case Some((key, value :: values :: values1)) => tref.update(_ + (key -> ::(values, values1))).as(Some(value))
+      }
+    }
+
+  /**
+   * Collects all values into a list.
+   */
+  def toList: USTM[List[V]] =
```

Comment:
just curious as to why providing a list collection is preferable over a vector.

Reply:
Actually, I'd go with chunks here from the beginning.